src/NosCore.Controllers/NpcPacketController.cs

```diff
@@ -135,10 +155,144 @@ public void Shopping(ShoppingPacket shoppingPacket)
+            {
+                Session.SendPacket(new MsgPacket
+                {
+                    Message = Language.Instance.GetMessageFromKey(LanguageKey.SHOP_NEAR_PORTAL,
+                        Session.Account.Language),
+                    Type = 0
+                });
+                return;
+            }
+
+            if (Session.Character.Group != null && Session.Character.Group?.Type != GroupType.Group)
+            {
+                Session.SendPacket(new MsgPacket
+                {
+                    Message = Language.Instance.GetMessageFromKey(LanguageKey.SHOP_NOT_ALLOWED_IN_RAID,
+                        Session.Account.Language),
+                    Type = 0
+                });
+                return;
+            }
+
+            if (!Session.Character.MapInstance.ShopAllowed)
+            {
+                Session.SendPacket(new MsgPacket
+                {
+                    Message = Language.Instance.GetMessageFromKey(LanguageKey.SHOP_NOT_ALLOWED,
+                        Session.Account.Language),
+                    Type = 0
+                });
+                return;
+            }
+
+            switch (mShopPacket.Type)
+            {
+                case CreateShopPacketType.Open:
+                    Session.Character.Shop = new Shop();
+                    short shopSlot = -1;
+                    foreach (var item in mShopPacket.ItemList)
+                    {
+                        shopSlot++;
+                        if (item.Amount == 0)
+                        {
+                            continue;
+                        }
+
+                        var inv = Session.Character.Inventory.LoadBySlotAndType<IItemInstance>(item.Slot, item.Type);
+                        if (inv == null)
+                        {
+                            //log
+                            continue;
+                        }
+                        if (inv.Amount < item.Amount)
+                        {
+                            //todo log
+                            return;
+                        }
+
+                        if (!inv.Item.IsTradable || inv.BoundCharacterId != null)
+                        {
+                            Session.SendPacket(Session.Character.GenerateSay(
+                                Language.Instance.GetMessageFromKey(LanguageKey.SHOP_ONLY_TRADABLE_ITEMS, Session.Account.Language),
+                                SayColorType.Yellow));
+                            Session.SendPacket(new ShopEndPacket { Type = 0 });
+                            return;
+                        }
+
+                        Session.Character.Shop.ShopItems.TryAdd(shopSlot,
+                            new ShopItem
+                            {
+                                Amount = item.Amount,
+                                Price = item.Price,
+                                Slot = shopSlot,
+                                Type = 0,
```

Comment:
Same here

Reply:
there is no type here it's the shop id. A character can only have 1 shop so 0